### PR TITLE
feat(metrics): Add metrics consumer for billing [INGEST-1609]

### PR DESCRIPTION
**Pull request by iker-barriocanal**
Add a consumer that reads from `snuba-metrics` (the topic the metrics indexer produces to), and calls `track_outcome` for processed transactions.

`d:transactions/duration@millisecond` is the one metric that we unconditionally extract for every transaction, so the number of elements in the bucket value corresponds to the number of transactions that contributed to the bucket of this metric. For each bucket, report this length as the `quantity` of the tracked outcome.

Future work:

Replace the usage of `track_outcome` with a streaming consumer that awaits delivery of outcomes before committing offsets.

### CODE REVIEW

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (iker-barriocanal) [inline src/sentry/ingest/billing_metrics_consumer.py:36]: Are there any other configuration parameters I should consider?

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (iker-barriocanal) [inline src/sentry/ingest/billing_metrics_consumer.py:95]: I assume calling this function here, instead of duplicating the logic, isn't a concern in terms of performance and scalability.

Comment (lynnagara) [reply src/sentry/ingest/billing_metrics_consumer.py:95]: What does this function do? Keep in mind this is blocking on the submit method (if it's something that can take a while). This way only one message is being processed at a time and the next message won't start until this function is done. Is that ok?

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (lynnagara) [inline src/sentry/ingest/billing_metrics_consumer.py:108]: The offset to be committed is always the next one after the current message's offset
```suggestion
        self.__ready_to_commit[message.partition] = Position(message.next_offset, message.timestamp)
```

**Comment by onewland** [comment]
It seems fine to me but I'm not clear on what `track_outcome` does underneath the hood.

If it's more efficient or supported, you might want to accumulate a batch in the `submit` method and then do an expensive `track_outcome[s]` on that batch in the `join`.

But if that's not possible I think what you're doing makes sense in principle

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (lynnagara) [reply src/sentry/ingest/billing_metrics_consumer.py:36]: Mind using this function to build the config? https://github.com/getsentry/arroyo/blob/main/arroyo/backends/kafka/configuration.py#L69 

Arroyo auto offset reset is a little bit different to normal and it allows us to also add the --strict-offset-reset later to our consumers.

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (iker-barriocanal) [reply src/sentry/ingest/billing_metrics_consumer.py:95]: It's this function:
https://github.com/getsentry/sentry/blob/87530245448936cea77125c521ece47ee51f9d78/src/sentry/utils/outcomes.py#L39
The function is not heavy computation-wise, so I discarded the alternative to perform the computation asynchronously.

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (iker-barriocanal) [inline src/sentry/ingest/billing_metrics_consumer.py:34]: @lynnagara I'm not sure this is the right way to do this. If this is not the right way, let me know what the best way is.

File: mypy.ini
Comment (jjbayer) [inline mypy.ini:60]: Should we move this to `src/sentry/sentry_metrics`, which is where the indexer consumer lives?

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (jjbayer) [reply src/sentry/ingest/billing_metrics_consumer.py:131]: `track_outcome` calls `KafkaPublisher.publish`, which itself sends a Kafka message asynchronously, so I _guess_ it's OK?

https://github.com/getsentry/sentry/blob/87530245448936cea77125c521ece47ee51f9d78/src/sentry/utils/outcomes.py#L108

https://github.com/getsentry/sentry/blob/58dbd38e3fa0edc67a19dda3182b9c46fdc126ca/src/sentry/utils/pubsub.py#L74-L75

@lynnagara what would be a better place to call this function? In `self.poll()`, maybe?

File: src/sentry/testutils/skips.py
Comment (jjbayer) [inline src/sentry/testutils/skips.py:69]: This is an unrelated change that seems to be related to the pytest update. Not sure why it does not fail tests on master.

File: tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py
Comment (jjbayer) [inline tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py:94]: For some reason I do not understand, this tests takes ~40 iterations = 40 seconds on average to complete, even though the produced metrics messages show up in [kcat](https://github.com/edenhill/kcat) right away.

File: mypy.ini
Comment (jjbayer) [reply mypy.ini:60]: Decided with @iker-barriocanal to leave it here for now.

File: tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py
Comment (jjbayer) [reply tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py:95]: This happened on subsequent test runs because I used the same group ID in every run.

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (jjbayer) [inline src/sentry/ingest/billing_metrics_consumer.py:107]: @lynnagara is it OK to commit on every poll, or do we need batched or throttled commits, like in 

https://github.com/getsentry/sentry/blob/6da3b229f74e1fe82b30d775c112e0490a41f22b/src/sentry/replays/consumers/recording/process_recording.py#L249-L256

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (iker-barriocanal) [reply src/sentry/ingest/billing_metrics_consumer.py:136]: cc @onewland, this should be cheap.

**Comment by iker-barriocanal** [comment]
The PR looks good to me, but I'm biased for having worked on it.

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (fpacifici) [reply src/sentry/ingest/billing_metrics_consumer.py:136]: Looking at `track_outcome`, it seems that function uses the legacy publisher to produce messages. That producer does not guarantee delivery as it does not wait for the callback. So you can loose outcomes, specifically in case of errors or when the consumer is shut down.
Is that acceptable ? 
If not please use the [arroyo producer](https://github.com/getsentry/arroyo/blob/main/arroyo/backends/kafka/consumer.py#L676) instead (that returns a future) so you can wait for the future result periodically and commit only when you have confirmation the message has been produced to Kafka.

File: tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py
Comment (fpacifici) [inline tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py:79]: I think this is going to be a fragile test as there is a waiting time every time you subscribe to a topic before you start receiving messages that depends on how long kafka takes to rebalance and assign you partitions. So 100 may be not enough in a flaky way.
Since you are mocking all the producer logic I would suggest you turn this into a test that does not depend on Kafka to be running.
Arroyo provides you three ways to test your consumer:
- unit test the processing strategy: This means instantiate `BillingTxCountMetricConsumerStrategy` with the factory you created and routinely call `submit` and `poll`. So you test your logic without the need of the entire Kafka scaffolding.
- test your consumer with the [arroyo in memory broker/consumer](https://github.com/getsentry/arroyo/blob/main/arroyo/backends/local/backend.py#L130). This is a fake of the Kafka consumer for unit tests so the response time is predictable.
- Or you test with the real Kafka but this should be used only when you want to run integration tests.

File: src/sentry/runner/commands/run.py
Comment (fpacifici) [inline src/sentry/runner/commands/run.py:663]: This decorator allows the client to specify `max-batch-size` and `max-batch-time-ms`.  That generally tells you how often you are supposed to commit.

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (fpacifici) [reply src/sentry/ingest/billing_metrics_consumer.py:107]: See https://github.com/getsentry/sentry/blob/f94d5af03ed666db8b5ad3898641bdb785ee17e4/src/sentry/runner/commands/run.py#L452-L466 to decide how often to commit.

File: src/sentry/runner/commands/run.py
Comment (fpacifici) [inline src/sentry/runner/commands/run.py:665]: You should not need to pass a topic name in the CLI and it is generally a bad practice.
You should not need it because, if you want to override the topic name, `batch_kafka_options` already provides you a [force-topic CLI argument](https://github.com/getsentry/sentry/blob/f94d5af03ed666db8b5ad3898641bdb785ee17e4/src/sentry/runner/commands/run.py#L476-L490).
It is a bad practice because it bypasses the system that maps topics to cluster in the settings. Topics are defined as abstract entities here: https://github.com/getsentry/sentry/blob/master/src/sentry/conf/server.py#L2442-L2472 then mapped to clusters here https://github.com/getsentry/sentry/blob/master/src/sentry/conf/server.py#L2489. The idea is that you could override topic names per environment via settings. That is horribly broken because how settings work in Sentry, still if you expect the topic name to be always passed in, you would be skipping all the logic that maps topic name to cluster which is even worse.

So the right approach is to remove this field and rely on the constant from server.py as a topic name. Also use the force-topic field if, for some reason, in some specific environment there is a reason to override the topic. Should very rarely happen

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (jjbayer) [reply src/sentry/ingest/billing_metrics_consumer.py:145]: > So you can loose outcomes, specifically in case of errors or when the consumer is shut down.
Is that acceptable ?

From my current understanding, it is acceptable, but I will let @jan-auer make the final call.

File: src/sentry/runner/commands/run.py
Comment (jjbayer) [reply src/sentry/runner/commands/run.py:663]: Thanks, now used by the strategy.

File: tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py
Comment (jjbayer) [reply tests/sentry/ingest/billing_metrics_consumer/test_billing_metrics_consumer_kafka.py]: Converted the test to a unit test calling `submit` and `poll` on the strategy repeatedly.

File: src/sentry/runner/commands/run.py
Comment (jjbayer) [reply src/sentry/runner/commands/run.py]: 👍 I removed the option.

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (jan-auer) [inline src/sentry/ingest/billing_metrics_consumer.py:81]: Since this isn't true for all buckets observed, could we leave that untyped and introduce an explicit check, or does that increase overheads too much?

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (jan-auer) [reply src/sentry/ingest/billing_metrics_consumer.py:145]: Let's go with this implementation to unblock other work, and move this to the arroyo producer in an immediate follow-up. We definitely want delivery guarantees.

**Review by jan-auer — APPROVED** [review]
This consumer is transaction-specific, so it should be marked with the transactions/performance cogs category on deployment. Even if it runs over buckets from other categories (i.e. sessions), the reason it exists is transactions, and thus that's where the cost should be attributed to. cc @oioki

File: src/sentry/ingest/billing_metrics_consumer.py
Comment (jjbayer) [reply src/sentry/ingest/billing_metrics_consumer.py]: Done.

**Pull request merged**
